functional_programming_tests/test/with_dependency_injection.cpp

```diff
@@ -0,0 +1,528 @@
+
+  for (size_t idx = 0; idx < (msg.data.size() - 1); idx += 2) {
+    path.push_back({msg.data[idx], msg.data[idx + 1]});
+  }
+
+  return path;
+}
+
+/**
+ * @brief Create a sample cost map in a request message
+ * @return A shared pointer to the request message with the cost map
+ */
+std::shared_ptr<example_srvs::srv::SetMap::Request> make_costmap() {
+  auto const request = std::make_shared<example_srvs::srv::SetMap::Request>();
+
+  request->map = std_msgs::msg::UInt8MultiArray();
+
+  request->map.layout.dim.resize(3, std_msgs::msg::MultiArrayDimension());
+
+  request->map.layout.dim[0].label = "rows";
+  request->map.layout.dim[0].size = 8;
+  request->map.layout.dim[0].stride = 64;
+
+  request->map.layout.dim[1].label = "columns";
+  request->map.layout.dim[1].size = 8;
+  request->map.layout.dim[1].stride = 1;
+
+  request->map.layout.dim[2].label = "channel";
+  request->map.layout.dim[2].size = 1;
+  request->map.layout.dim[2].stride = 1;
+
+  request->map.data = {0, 0, 0, 0, 0, 0, 0, 0,  //
+                       0, 0, 0, 1, 0, 0, 0, 0,  //
+                       0, 0, 0, 1, 0, 0, 0, 0,  //
+                       0, 0, 1, 1, 1, 0, 0, 0,  //
+                       0, 0, 1, 0, 1, 1, 0, 0,  //
+                       0, 0, 1, 0, 0, 0, 0, 0,  //
+                       0, 0, 0, 0, 0, 0, 0, 0,  //
+                       0, 0, 0, 0, 0, 0, 0, 0};
+  return request;
+}
+
+// Create mock middleware handle
+struct MockMiddlewareHandle : public PathGenerator::MiddlewareHandle {
+  MOCK_METHOD(void, register_set_map_service, (SetMapServiceCallback),
+              (override));
+  MOCK_METHOD(void, register_generate_path_service,
+              (GeneratePathServiceCallback), (override));
+  MOCK_METHOD(void, log_info, (std::string const&), (override));
+  MOCK_METHOD(void, log_error, (std::string const&), (override));
+};
+
+TEST(PathGenerator, Ctor) {
+  // GIVEN a mock middleware handle
+  auto mw = std::make_unique<MockMiddlewareHandle>();
+
+  // THEN it should register the services
+  EXPECT_CALL(*mw, register_set_map_service(testing::_)).Times(1);
+  EXPECT_CALL(*mw, register_generate_path_service(testing::_)).Times(1);
+
+  // WHEN the path generator is constructed
+  auto const path_generator = PathGenerator{std::move(mw)};
+}
+
+TEST(PathGenerator, SetMap) {
+  // GIVEN a path generator
+  auto mw = std::make_unique<MockMiddlewareHandle>();
+  // Capture the callback so it can be called later
+  PathGenerator::MiddlewareHandle::SetMapServiceCallback callback;
+  ON_CALL(*mw, register_set_map_service(testing::_))
+      .WillByDefault(testing::SaveArg<0>(&callback));
+
+  auto const path_generator = PathGenerator{std::move(mw)};
+
+  // WHEN the set map service is called
+  auto const request = make_costmap();
+  auto response = std::make_shared<example_srvs::srv::SetMap::Response>();
+  callback(request, response);
+
+  // THEN the path generator should successfully set the map
+  EXPECT_TRUE(response->success.data);
+}
```

Comment:
Here, we have `callback`, which is the std::function type for the set map service callback, and when `register_set_map_service` is called by PathGenerator, save the callback that the PathGenerator registers, and then test the callback?

Reply:
Exactly. Perfect.
![](https://media.giphy.com/media/1236TCtX5dsGEo/giphy.gif)